parser/typecheck/typecheck.py

```diff
@@ -0,0 +1,227 @@
+
+    def _init(self):
+        self.ast = self.astgen.parse()
+        self.top_scope = Scope()
+
+    def run(self):
+        if self.top_scope:
+            return self.top_scope
+        self._init()
+        self.run_on_new_scope(self.ast.statements, curr_scope=self.top_scope)
+        return self.top_scope
+
+    def run_on_new_scope(self, block: list[AstNode], scope_stack: list[Scope] = None,
+                         curr_scope: Scope = None):
+        def enter_ident(n: AstIdent):
+            for s in scope_stack[::-1]:  # Inefficient, creates a copy!
+                if info := s.declared.get(n.id):
+                    curr_scope.used[n.id] = info
+                    return
+            raise self.err(f"Name '{n.id}' is not defined", n.region)
+
+        def enter_decl(n: AstDeclNode):
+            # Need semi-special logic here to prevent walking it walking
+            # the AstIdent that is currently being declared.
+            AstNode.walk_obj(n.value, walker)  # Don't walk `n.ident`
+            # Do this after walking (that is when the name is bound)
+            ident = n.ident.id
+            target_scope = curr_scope if n.scope == VarDeclScope.LET else self.top_scope
+            if ident in target_scope.declared:
+                raise self.err("Variable already declared", n.region)
+            target_scope.declared[ident] = NameInfo(target_scope, ident, (
+                ValType() if n.type == VarDeclType.VARIABLE else ListType()))
+            return True
+
+        def enter_fn_decl(fn: AstDefine):
+            ident = fn.ident.id
+            if ident in curr_scope.declared:
+                raise self.err("Function already declared", fn.ident.region)
+            subscope = Scope()
+            params: list[ParamInfo] = []
+            for tp, param in fn.params:
+                if tp.id not in PARAM_TYPES:
+                    raise self.err("Unknown parameter type", tp.region)
+                if param.id in subscope.declared:
+                    raise self.err("There is already a parameter of this name", param.region)
+                tp = BoolType() if param.id == 'bool' else ValType()
+                subscope.declared[param.id] = NameInfo(subscope, param.id, tp, is_param=True)
+                params.append(ParamInfo(param.id, tp))
+            curr_scope.declared[ident] = info = FuncInfo.from_param_info(
+                curr_scope, ident, params, VoidType(), subscope)
+            inner_funcs.append((info, fn))  # Store funcs for later walking
+            # Skip walking body, only walk inner after collecting all declared
+            #  variables in outer scope so function can use all variables
+            #  declared in outer scope - even the ones declared below it)
+            return True
+
+        curr_scope = curr_scope or Scope()
+        # Can't use `or` because need to preserve reference if arg is `[]`
+        scope_stack = scope_stack if scope_stack is not None else []
```

Comment:
This comment here makes no sense?! Why do we need to preserve the reference?

Reply:
(We will just do one push, then one pop so we just return the same empty list - which is a bit useless to preserve the ref)